azure-kusto-data/azure/kusto/data/_token_providers.py

```diff
@@ -370,6 +370,43 @@ def _get_token_from_cache_impl(self) -> dict:
         return self._valid_token_or_none(token)
 
 
+class InteractiveLoginTokenProvider(TokenProviderBase):
+    """ Acquire a token from MSAL with Device Login flow """
+
+    def __init__(self, kusto_uri: str, authority_uri: str, login_hint: str = None, domain_hint: str = None):
+        super().__init__(kusto_uri)
+        self._msal_client = None
+        self._auth = authority_uri
+        self._login_hint = login_hint
+        self._domain_hint = domain_hint
+        self._account = None
+
+    @staticmethod
+    def name() -> str:
+        return "DeviceLoginTokenProvider"
+
+    def context(self) -> dict:
+        return {"authority": self._auth, "client_id": self._cloud_info.kusto_client_app_id}
+
+    def _init_impl(self):
+        self._msal_client = PublicClientApplication(client_id=self._cloud_info.kusto_client_app_id, authority=self._auth)
+
+    def _get_token_impl(self) -> dict:
+        token = self._msal_client.acquire_token_interactive(
+            scopes=self._scopes, prompt="select_account", login_hint=self._login_hint, domain_hint=self._domain_hint
+        )
+        return self._valid_token_or_throw(token)
+
+    def _get_token_from_cache_impl(self) -> dict:
+        account = None
+        accounts = self._msal_client.get_accounts(self._login_hint)
+        if len(accounts) > 0:
+            account = accounts[0]
+
+        token = self._msal_client.acquire_token_silent(scopes=self._scopes, account=account)

```

Comment:
Are you sure that acquire_token_silent doesn't throw when account is None?
Maybe worth checking before and return None

Reply:
On the first pass before the prompt is called there are no accounts.
Token acquisition always starts with the cache and then goes to the actual API, so this happens every time.